Handle ROMS runtime settings without template files

dafyddstephenson opened the pull request:
- [x] Closes #244 
- [x] Tests added
- [x] Tests passing
- [x] Full type hint coverage
- [x] Changes are documented in `docs/releases.rst`
- [x] New functions/methods are listed in `api.rst`
- [x] New functionality has documentation

Notes, 20250416: Have defined the class and `to_` and `from_file` methods, and checked with ROMS using every configuration in `Examples` plus one of my own - the roundtrip produces bitwise identical output for those cases, but not all entries are covered.

Commit:
Create ROMSInputSettings class with 'from_file' and 'to_file' methods

Commit:
add docstrings to 'read_inp.py', correct types for MY_bak_mixing, tidy up unused code

Commit:
Merge branch 'main' into roms_input_settings

Commit:
- Rename ROMSInputSettings -> ROMSRuntimeSettings
- Update ROMSInputDataset.partition to write to same directory as input files
- Update ROMSSimulation to use ROMSRuntimeSettings rather than string replacements in template files
- Remove modified_files attribute on AdditionalCode

Commit:
update str and repr fns and docstrings

Commit:
update blueprints for integration tests

Commit:
begin updating unit tests

Commit:
update existing unit tests

Commit:
get roms/simulation.py unit test coverage back up to 100%

Commit:
begin adding unit tests for ROMSRuntimeSettings

Commit:
get up to 100% unit test coverage for ROMSRuntimeSettings

Commit:
update docstrings for test_roms_simulation.py, fix typos in test_read_inp.py

Commit:
add docstrings to ROMSRuntimeSettings unit tests, add more typehints, tidy up

Commit:
update api.rst and releases.rst

dafyddstephenson commented:
@ScottEilerman this isn't ready for merge (want to get @ankona 's logging changes in first and then deal with any conflicts) but it is probably(?) ready for review. Let me know if this matches what you were anticipating

ScottEilerman commented on cstar/roms/simulation.py:
```diff
@@ -1013,15 +995,15 @@ def tree(self):
+        └─── ROMS
+             ├── runtime_code
+             │   ├── namelist_file.in
+             │   └──  another_file.in
```
I think a second `.in` is forbidden now, right? Based on one of the errors thrown when auto-detecting the .in.

dafyddstephenson replied:
f8038f6ba39002f0449e8e127bdaf8a6deefa6b1

Commit:
Merge branch 'main' into roms_input_settings

Commit:
re-run tutorial notebooks

Commit:
TMP switch to SE's read_inp and some tidying/annotation

Commit:
merge serializers into single serializer

Commit:
- Add subclasses for /every/ ROMS input file section
- Add 'from_lines' classmethod to RomsSection
- Simplify logic in 'from_file' to read file to lines and populate ROMSRuntimeSection using lines

Commit:
rename OneEntryRomsSection -> SingleEntryRomsSection

dafyddstephenson commented:
@ScottEilerman still needs some docstring and test tweaks but here's what I got starting from your attempt - if this is looking ballpark I'll go ahead and tidy up. Sorry this turned into a long one but was definitely worth it imho!

Commit:
add pydantic to dependencies

ScottEilerman committed:
Add a validator to handle init from file more cleanly

Commit:
update simulation.py to use new RomsRuntimeSettings structure

ScottEilerman committed:
Run pre-commit, revert some minor changes

dafyddstephenson committed:
Merge pull request #59 from ScottEilerman/validate-from-lines

Add a validator to handle init from file more cleanly

Commit:
Improvements and bugfixes
* Renames for clarity and explicitness:
- rename read_inp.py -> runtime_settings.py
- rename test_read_inp.py -> test_roms_runtime_settings.py
- rename RomsSection -> ROMSRuntimeSettingsSection
- rename SingleEntryRomsSection -> SingleEntryROMSRuntimeSettingsSection
- rename InitialBlock -> InitialConditions
- rename ForcingBlock -> Forcing
- rename MarblBGC -> MARBLBiogeochemistry
* Remove unused keys BottomDrag.Cdb_{min,max}
* Correct MYBakMixing class to be multi-entry
* Address case-insensitive section names in from_file with lower()
* Update ROMSSimulation.roms_runtime_settings property def
* Update example '.in' file in test fixtures
* Update example ROMSRuntimeSettings instance in test_roms_runtime_settings.py
* Fix unit tests

Commit:
Merge branch 'main' into roms_input_settings

Commit:
update docstrings and add validator for SingleEntryROMSRuntimeSettingsSection

Commit:
Add validator to simplify SingleEntryROMSRuntimeSettings initialisation; Add ROMSInputDataset.path_for_roms property to take partitioned files and create a path ROMS can read

Commit:
delete unused check in path_for_roms, begin adding unit tests

ScottEilerman commented on cstar/roms/runtime_settings.py:
```diff
@@ -136,10 +138,10 @@ def from_lines(cls, lines: Optional[list[str]]) -> Optional["RomsSection"]:
+            InitialConditions(nrrec=1, ininame=Path("input_datasets/roms_ini.nc"))
         """
-        if lines is None:
+        if (lines is None) or (len(lines) == 0):
```
nit: would `if not lines` suffice here?

dafyddstephenson replied:
f8038f6ba39002f0449e8e127bdaf8a6deefa6b1

ScottEilerman commented on cstar/roms/simulation.py:
```diff
@@ -625,37 +626,33 @@ def roms_runtime_settings(self) -> ROMSRuntimeSettings:
+        if self.model_grid:
+            simulation_runtime_settings.grid = cstar.roms.runtime_settings.Grid(
+                grid=self.model_grid.path_for_roms[0]
+            )
```
nit: I think we can use the shortcut here instead of making the Grid() object explicitly, right?

dafyddstephenson replied:
I don't think so - at least I tried and it didn't work 😬 just assigns the `grid` attribute as a `Path` instance, which makes sense. I think it's only at `__init__` of the whole `ROMSRuntimeSettings` instance that our changes are effective - re-assigning attributes afterwards skips over them

ScottEilerman replied:
Hmm interesting. I thought we had a counterexample but I'm not seeing it now, so you may be right. I was going to say, it could be an issue with `Path` in particular -- usually after instantiation these things end up as `PosixPath` and the type-checking from the annotation might not register correctly, the way we did it.

It might be worth a quick try of changing the type hint to `Path | PosixPath` and see if it suddenly works, but if not, it's not a big deal at all and we can move on.

ScottEilerman replied:
Confirming that this is about pydantic not re-validating on attribute assignment. There's a setting for that, but I think that will break some of the None handling, so I'm going to leave it as-is for now.

ScottEilerman reviewed (approved):
A few small things but otherwise this is looking great!

Commit:
bugfixes and unit test additions

Commit:
100% line coverage for roms/runtime_settings.py

Commit:
add docstrings for test_roms_runtime_settings.py

Commit:
responding to SE's PR review

dafyddstephenson commented:
Notes to self:
The roundtrip `ROMSRuntimeSettings.from_file(f).to_file(f2)` still fails for:
- `$ROMS_ROOT/Examples/Rivers_ana/river_ana.in` (no initial condition file)
- `$ROMS_ROOT/Examples/Pipes_ana/pipes_ana.in` (no initial condition file)
- `$ROMS_ROOT/Examples/Filament/Filament.in` (no forcing files)

Converting to draft until fixed

ankona commented on ci/environment.yml:
```diff
@@ -10,6 +10,7 @@ dependencies:
   - mpich
   - xarray
   - netCDF4
+  - pydantic
```
do you want to pin to >= 2 or anything here?

dafyddstephenson replied:
6a4845fa3741d98bbb187e546c02f7846c2d97f3

ankona commented on cstar/roms/input_dataset.py:
```diff
@@ -432,6 +432,29 @@ def _update_partitioning_attribute(
         }
 
+    @property
+    def path_for_roms(self) -> list[Path]:
```
Do you always want to keep a "complete list?" Would a regex here be more compact and capture what you need?

ankona commented on cstar/roms/input_dataset.py:
```diff
@@ -432,6 +432,29 @@ def _update_partitioning_attribute(
+            zero_files = [f for f in self.partitioning.files if zero_str in str(f)]
+            return [Path(str(f).replace(zero_str, ".nc")) for f in zero_files]
+
+        else:
```
else is unnecessary after a return. consider removing.

dafyddstephenson replied:
b4334d1a888626b8011f6a498b21b7e07981f226

ankona commented on cstar/roms/input_dataset.py:
```diff
@@ -432,6 +432,29 @@ def _update_partitioning_attribute(
+        else:
+            raise FileNotFoundError(
+                "ROMS requires files to be partitioned for use"
+                "call ROMSInputDataset.partition() or ROMSSimulation.pre_run() "
```
looks like you might have a couple chars w/no whitespace after concatenation - `usecall`

dafyddstephenson replied:
6edec131c9e682622234f87adb34f5a4fd64c11e

ankona commented on cstar/roms/input_dataset.py:
```diff
@@ -432,6 +432,29 @@ def _update_partitioning_attribute(
+        """
+        if self.partitioning is not None:
+            ndigits = len(str(self.partitioning.np_xi * self.partitioning.np_eta))
+            zero_str = "." + "0" * ndigits + ".nc"
```
Consider using a regex instead for readability and to simplify your renaming. Something like this lets you grab the filenames that match.

given:
```
>>> filenames = "\n".join(os.listdir(p))
>>> print(filenames)
foo.5.nc
foo.1.nc
data_3.002.nc
data_1.002.nc
foo.000nc
foo.4.nc
data_2.000.nc
data_2.001.nc
foo.000.nc
data_2.002.nc
foo.8.nc
foo.3.nc
data_1.001.nc
data_3.000.nc
foo.7.nc
foo.6.nc
foo.2.nc
data_1.000.nc
data_3.001.nc
```

this should accomplish the same with less looping:

```py
>>> matches = re.findall(f"(.*)\\.{'0' * ndigits}\\.nc", filenames)
>>> matches
['data_2', 'foo', 'data_3', 'data_1']
>>> targets = [f"{x}.nc" for x in matches]
>>> targets
['data_2.nc', 'foo.nc', 'data_3.nc', 'data_1.nc']
```

dafyddstephenson replied:
This is neat, but I don't think I could avoid a second loop here as `self.partitioning.files` is a list of `Path` objects that would have to be converted to `str` iteratively first. This may be worth revisiting when we have thousands of files if it seems slow, but for now I think I'll keep it as is.

ankona commented on cstar/roms/input_dataset.py:
```diff
@@ -432,6 +432,29 @@ def _update_partitioning_attribute(
+            ndigits = len(str(self.partitioning.np_xi * self.partitioning.np_eta))
+            zero_str = "." + "0" * ndigits + ".nc"
+            zero_files = [f for f in self.partitioning.files if zero_str in str(f)]
+            return [Path(str(f).replace(zero_str, ".nc")) for f in zero_files]
```
Do you need to perform any sanity checking here to make sure you have a contiguous set of files without gaps?

As-is, is it ok to return `foo.nc` if the directory contained `[foo.000.nc, foo.001.nc, foo.099.nc]` and not notice the missing 002-098 parts?

dafyddstephenson replied:
I think this should be raised as an issue - you're right, but this isn't the line for that check and this isn't the PR to add it

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,756 @@
+
+def _format_list_of_floats(float_list: list[float]) -> str:
+    joiner = " "
+    return joiner.join(_format_float(x) for x in float_list)
```
I'm not sure this method has any value, given the hardcoded delimiter. Parameterizing the delimiter at least makes it re-usable.

```
def _format_list_of_floats(floats: list[float], delimiter) -> str:
    return delimiter.join(_format_float(x) for x in floats)
```

But why not just put a single line in place of `_format_list_of_floats` and avoid the extra method, altogether. You can add a constant for your delimiter if you desire:

```
LIST_DELIMITER = ""
...
...
def some_other_method():
    ...
    formatted_floats = LIST_DELIMITER.join(_format_float(x) for x in float_list)
```

and... after reading the additional formatting methods, it seems like you really just want to use `map`?

```py
>>> items = [1, 2, 3, 4, 5, 6]
>>> formatted = list(map(lambda x: f"{x:.2f}", items))
>>> formatted
['1.00', '2.00', '3.00', '4.00', '5.00', '6.00']
>>>
>>>
>>> formatted = list(map(str, items))
>>> formatted
['1', '2', '3', '4', '5', '6']
>>>
>>> # or use your own formatting method...
>>> formatted = list(map(_format_float, items))

```

you can get rid of `_format_path`, `_format_other` w/`list(map(str, items))`

dafyddstephenson replied:
Anywhere C-Star interfaces with ROMS is a recipe for chaos - this module is deliberately _extremely_ explicit (and contains methods that aren't even used) to try and foster some sort of structure/readability/maintainability and remove some of the "why the hell are they doing that" factor that inevitably arises when just doing the minimal thing that makes ROMS work. 

I'm not sure where the suggestion to `map` fits in, but the formatting methods are supposed to return single strings (however trivially that's achieved) that are ready to go in the bewilderingly structured `roms.in` file (your examples return lists of strings?)

ankona replied:
`map` applies a function to every member of a collection. Creating functions that just perform iteration is unnecessary.

This method can be avoided:

```py
def _format_list_of_paths(
    path_list: list[Path], multi_line: Optional[bool] = False
) -> str:
    joiner = "\n    " if multi_line else " "
    return joiner.join(_format_path(x) for x in path_list)
```

with:

```py
DELIMITER.join(map(str, the_collection))
```

and this method can be avoided:

```py
def _format_list_of_floats(float_list: list[float]) -> str:
    joiner = " "
    return joiner.join(_format_float(x) for x in float_list)
```

with:

```py
DELIMITER.join(map(str, the_other_list))
```

and, this method can be avoided:

```py
def _format_list_of_other(other_list: list[str | int]) -> str:
    joiner = " "
    return joiner.join(_format_other(x) for x in other_list)
```

with:

```py
DELIMITER.join(map(str, the_last_list))
```

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,756 @@
+        # from the roms.in file, and try to parse it as such. if that fails, or if it's not a list
+        # when it comes in, do the usual init process.
+
+        if isinstance(data, list) and all([isinstance(v, str) for v in data]):
```
Consider inverting this check to reduce nesting. For example...

```py
if not isinstance(list):
  raise ValueError("Not a list")

# now i know it is a list
if not all(isinstance(v, str) for v in data):  # avoid [] brackets to do lazy load
  raise ValueError("You didn't give me strings")

try:
  ...
except:
  ...
```

Reducing nesting almost always makes the code easier to follow.

dafyddstephenson replied:
Appreciate the focus on readability but I think I disagree here - the single statement reads left to right quite intuitively, while split following the suggestion above it would look like:

```
if not isinstance(data, list):
    return handler(data)

if not all(isinstance(v, str) for v in data):
    return handler(data)

try:
    return cls.from_lines(data)
except Exception:
    return handler(data)
```
Good point about the square brackets though. Fixed in ed7a9ec40cd3769b5d4b0dfa16ecfca26415c67f

ankona replied:
This has the exact same behavior as the current code (though it may not pass a type-checker):

```python
def validate_from_lines(cls, data, handler: ModelWrapValidatorHandler):
      try:
          return cls.from_lines(data)
      except Exception:
          pass
      return handler(data)
```

If we're not actually going to perform correct validation, why do it at all?

If we don't care about reporting individual issues, inverting the control block would give us:

```py
def validate_from_lines(cls, data, handler: ModelWrapValidatorHandler):
     if not isinstance(data, list) or not all([isinstance(v, str) for v in data]):
          return handler(data)

      try:
          return cls.from_lines(data)
      except Exception:
          raise ValueError("improperly formatted lists")

```

or, if you're ok with whatever validation comes out of the class constructor:


```py
def validate_from_lines(cls, data, handler: ModelWrapValidatorHandler):
     if not isinstance(data, list) or not all([isinstance(v, str) for v in data]):
          return handler(data)

      return cls.from_lines(data)lists")

```

I think maybe this validator is implemented this way so it can run on the subclasses that don't contain lists (e.g. Title). If so, it's a poor choice. We're hiding failures to avoid having a validator at a higher level in the class hierarchy.

Why not place this validator (sans the always-on return handler(data)) on a `MultiEntryROMSRuntimeSettingsSection` class so it isn't inherited by `SingleEntryROMSRuntimeSettingsSection`? Of course, this adds a layer to the hierarchy, but it's definitely cleaner and doesn't have to swallow failures.

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,756 @@
+
+    @property
+    def value_joiner(self):
+        return "\n    " if self.multi_line else "    "
```
This appears to be a duplicate of `_format_list_of_paths`. You should probably get rid of one

dafyddstephenson replied:
this line in particular is duplicated but this is more coincidence - the same thing is needed twice in two different contexts and could (technically) change in each independently

ankona replied:
This strategy fails the `DRY` principle, and uses a poorly named method, to boot. If you were to implement `_format_list_of_paths` as:

```python
  joiner = self.value_joiner
  return joiner.join(...)
```

You would avoid having the same formatting specified in both places.

Why not do the above suggestion, and add in a rename to clarify what the method returns (prepending `_` if it really shouldn't be "public"):

```python
~~def value_joiner(...):~~
def _delimiter(...):
   ...
```

resulting in:
```python
def _format_list_of_paths(...):
  return self._delimiter.join(...)
```

ScottEilerman replied:
Did a version of this. It got a little trickier because one space was used for lists of values that are all part of a single "value", while multiple spaces are used in between logical "values". This was probably not worth the time to refactor, but it's done now!

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,756 @@
+
+        section_header = f"{self.section_name}: {' '.join(data.keys())}\n"
+        # Build the serialized string
+        string = ""
```
how about `document` or `serialized` instead of the var name `string`? slightly more descriptive and enables removing an unnecessary comment

dafyddstephenson replied:
012c83d

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,756 @@
+        Climatology file path
+    """
+
+    model_config = {"populate_by_name": True}
```
should this be defined above the comment block w/the rest of the member vars for consistency?

dafyddstephenson replied:
8c8d5bf

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,756 @@
+        return sections
+
+    @classmethod
+    def from_file(cls, filepath: Path | str) -> "ROMSRuntimeSettings":
```
Consider `PathLike` instead

dafyddstephenson replied:
7892e32

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,756 @@
+        sections = {k.lower(): v for k, v in sections.items()}
+        if not all(
+            key in sections.keys()
+            for key in ["title", "time_stepping", "bottom_drag", "output_root_name"]
```
consider a set operation for clarity:

```
REQUIRED_FIELDS = {....}
...

common = set(sections.keys()).intersection(REQUIRED_FIELDS)
if not common.issuperset(REQUIRED_FIELDS):
  ...

dafyddstephenson replied:
6b382af

ankona commented on cstar/roms/simulation.py:
```diff
@@ -240,16 +251,59 @@ def __init__(
+
+        in_files = [f for f in self.runtime_code.files if f.endswith(".in")]
+        if len(in_files) != 1:
+            raise ValueError(
```
ValueError might be the wrong type to raise here. Consider RuntimeError

dafyddstephenson replied:
581b6a0

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,756 @@
+    @model_validator(mode="wrap")
+    @classmethod
+    def cast_to_obj(cls, data, handler: ModelWrapValidatorHandler):
+        """Allows a SingleEntryROMSRuntimeSettingsSection to be initialized with just
```
docstring to method name mismatch? is this really casting? maybe `def load_single`?

dafyddstephenson replied:
c40d415

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,756 @@
+            try:
+                return cls(**{cls.section_name: data})
+            except Exception:
+                pass
```
consider `log.debug` on cast failures for debugging

dafyddstephenson replied:
52071ab

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,756 @@
+        super().__init_subclass__(**kwargs)
+
+        # Get all annotated instance fields (ignore ClassVars like section_name)
+        field_names = [
```
Getting the field names in order looks like it would be better located on the base class of this item - `ROMSRuntimeSettingsSection`.

dafyddstephenson replied:
This seems re-usable so I think this suggestion would make sense but as we're in `__init_subclass__` access isn't as straightforward - it's also not re-used, so I think is more readable here.

ankona replied:
You're setting `key_order` here with `field_names` and repeating the same thing:

here...

```python
class TimeStepping(ROMSRuntimeSettingsSection):
    ntimes: int
    dt: int
    ndtfast: int
    ninfo: int

    section_name = "time_stepping"
    key_order = ["ntimes", "dt", "ndtfast", "ninfo"]
```

and here...

```python
class BottomDrag(ROMSRuntimeSettingsSection):
    rdrg: float
    rdrg2: float
    zob: float

    section_name = "bottom_drag"
    key_order = ["rdrg", "rdrg2", "zob"]
```

and here...

```python

class InitialConditions(ROMSRuntimeSettingsSection):
    nrrec: int
    ininame: Path

    section_name = "initial"
    multi_line = True
    key_order = ["nrrec", "ininame"]
```

and here...

```python

class Forcing(ROMSRuntimeSettingsSection):
    filenames: list[Path]

    section_name = "forcing"
    multi_line = True
    key_order = ["filenames"]
```

and here...

```python

class VerticalMixing(ROMSRuntimeSettingsSection):
    Akv_bak: float
    Akt_bak: list[float]
    section_name = "vertical_mixing"
    key_order = ["Akv_bak", "Akt_bak"]
```

and here...

```python

class MARBLBiogeochemistry(ROMSRuntimeSettingsSection):
    marbl_namelist_fname: Path
    marbl_tracer_list_fname: Path
    marbl_diag_list_fname: Path

    section_name = "MARBL_biogeochemistry"
    multi_line = True
    key_order = [
        "marbl_namelist_fname",
        "marbl_tracer_list_fname",
        "marbl_diag_list_fname",
    ]
```

and here...

```python

class SCoord(ROMSRuntimeSettingsSection):
    theta_s: float
    theta_b: float
    tcline: float

    section_name = "s-coord"
    key_order = ["theta_s", "theta_b", "tcline"]
```

If it was in the base class, you could set key order in the base class instead of repeating yourself.

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,756 @@
+            if get_origin(vartype) != ClassVar
+        ]
+
+        if len(field_names) != 1:
```
Consider writing a custom validator for this class instead that runs this `len==1` check to keep validation in the right place.

dafyddstephenson replied:
Thanks for the suggested consideration, consider it considered! But I think I'm going to prioritise my wellbeing over writing another validator for a good while...

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,756 @@
+        cls.key_order = field_names
+
+        # Set section_name = <field name> if not explicitly provided
+        if not hasattr(cls, "section_name"):
```
`section_name` is an attribute defined on the base class. I am not sure, but it seems like that means this is both unnecessary and not doing what i think it's meant to?

why not avoid hasattr?

```
if not self.section_name:
    self.section_name = self.field_names[0]
```

dafyddstephenson replied:
the suggestion will raise as it assumes `section_name` is defined - while `section_name` is annotated on the base class, it isn't defined there. The existing code works as intended, I believe this would raise an `AttributeError`

ankona replied:
It could be solved by setting a default:

```py
class ROMSRuntimeSettingsSection(BaseModel):
    section_name: ClassVar[str] = Field(default="")
```

and `__init_subclass__` can use:

```py
if not cls.section_name:
    cls.section_name = field_names[0]
```

which does feel wonky...

alternatively, it could be solved by implementing field names and the section name method in the base. This avoids having to find section name population in the `__subclass_init__` method and, i think, allows us to scrub `key_order`?

```python
class ROMSRuntimeSettingsSection(BaseModel):
    _section_name: ClassVar[str]
    multi_line: ClassVar[bool] = False

    @classmethod
    def _field_names(cls) -> list[str]:
        """Return the fields of the model."""
        return [
            name
            for name, vartype in cls.__annotations__.items()
            if get_origin(vartype) != ClassVar
        ]

    @classmethod
    def section_name(cls) -> str:
        """Get the section name for the ROMS runtime settings."""
        if not hasattr(cls, "_section_name"):
            all_fields = cls._field_names()
            cls._section_name = all_fields[0]
        return cls._section_name


class FooSection(ROMSRuntimeSettingsSection):
    x: int
    y: int
    z: int
    _section_name: ClassVar[str] = "FooStuff!"


class BarSection(ROMSRuntimeSettingsSection):
    xxx: int
    yyy: int
    zzz: int


class BazSection(ROMSRuntimeSettingsSection):
    b: int
    a: int
    c: int
    
def test_field_names_in_base():
    foo = FooSection(x=42, y=84, z=168)
    bar = BarSection(yyy=48, xxx=24, zzz=96)
    baz = BazSection(a=1, b=2, c=3)

    assert foo.section_name() == "FooStuff!"
    assert bar.section_name() == "xxx"
    assert baz.section_name() == "b"
```

ankona replied:
Here's another version that uses a mixin to encapsulate the section naming. Note, it's now clear that on all of these `XxxRuntimSettingSection` classes, making the property name `section_name` is unnecessary (since the type is a section.

Also notable, instead of writing my own `field_names` I used `list(cls.__pydantic_fields__.keys())`

```python

def field_names(cls) -> list[str]:
    """Return the fields of a model."""
    return list(cls.__pydantic_fields__.keys())


class TitledSection(BaseModel):
    """A section that must specify a name."""
    name: str

    def model_post_init(self, _context: Any) -> None:
        """Post-initialization to set the name."""
        if not hasattr(self, "name"):
            raise ValueError("name must be defined.")


class AutoTitledSection(BaseModel):
    """A section with an automatically generated name."""

    @computed_field
    def name(self) -> str:
        """Automatically set the name to the name of the first model field."""
        all_fields = field_names(self)
        return all_fields[0]


class ROMSRuntimeSettingsSection(BaseModel):
    multi_line: ClassVar[bool] = False


class FooSection(ROMSRuntimeSettingsSection, TitledSection):
    x: int
    y: int
    z: int
    name: str = "FooStuff!"


class BarSection(ROMSRuntimeSettingsSection, AutoTitledSection):
    xxx: int
    yyy: int
    zzz: int


class BazSection(ROMSRuntimeSettingsSection, AutoTitledSection):
    b: int
    a: int
    c: int


def test_field_names_in_base():
    foo = FooSection(x=42, y=84, z=168)
    bar = BarSection(yyy=48, xxx=24, zzz=96)
    baz = BazSection(a=1, b=2, c=3)

    assert foo.name == "FooStuff!"
    assert bar.name == "xxx"
    assert baz.name == "b"
```

though I liked the exercise, I have to wonder - if we're willing to hardcode all the lists for `key_order` in every subclass, why can't we just set the section name and avoid relying on auto-naming completely?

ScottEilerman replied:
section names have been eliminated in favor of auto-snake-casing and a " custom cases" lookup.

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,756 @@
+
+    @staticmethod
+    def _load_raw_sections(filepath: Path) -> dict:
+        filepath = Path(filepath)
```
unnecessary secondary cast to Path

dafyddstephenson replied:
68525c1

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,756 @@
+            if ":" in line and not line.startswith("!"):
+                section_name = line.split(":")[0].strip()
+                i += 1
+                section_lines = []
```
consider using `sections = defaultdict(list)` above and you can operate directly on the KV store:

```
sections[section_name].append(line)
```

dafyddstephenson replied:
Changing to `defaultdict` here re-opened a wound I spent all day yesterday closing so I'm going to take the advice and nod approvingly and then not implement it 😅

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,756 @@
+                i += 1
+                section_lines = []
+
+                while i < len(lines) and ":" not in lines[i]:
```
Something feels really fishy about the duplication between the outer while loop and the inner while loop.

Can't we flatten this and use a single loop instead of duplication of checks for `!` and `:` and the stripping, and the incrementing of the counter?

dafyddstephenson replied:
yeah that was ugly. Fixed
8135e2312e611c86710b9e2d6c69294eddae8fe3

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,756 @@
+    model_config = {"populate_by_name": True}
+
+    @staticmethod
+    def _load_raw_sections(filepath: Path) -> dict:
```
this looks like it can be tightened to a `dict[str,str]`

dafyddstephenson replied:
close! `dict[str,list[str]]`
8135e2312e611c86710b9e2d6c69294eddae8fe3

ankona commented on cstar/roms/simulation.py:
```diff
@@ -439,57 +513,168 @@ def codebases(self) -> list[ExternalCodeBase]:
+
+        forcing_paths: list[Path] = []
+
+        for source in forcing_sources:
```
consider conditionally inserting into `forcing_sources` to avoid the need for null checks everywhere that uses it.

e.g. 

```
if self.surface_forcing:
  forcing_sources.extend(self.surface_forcing)
...
```
this will get rid of the `if source is None` too. 

alternatively, pre-clean:

```
>>> x = [1, None, 2, None, 3, None]
>>> 
>>> forcing_sources = list(filter(lambda atom: atom is not None, x))
>>> forcing_sources
[1, 2, 3]
>>> for source in forcing_sources:
```

alternatively, use itertools...


```
all_sources = itertools.chain.from_iterable([self.tidal, self.river, self.surface, self.boundary, ...])
forcing_sources = list(filter(lambda atom: atom is not None, x))
```

dafyddstephenson replied:
good suggestion, cheers! 91884bc

ankona commented on cstar/roms/simulation.py:
```diff
@@ -1408,14 +1391,19 @@ def run(
-        n_time_steps = run_length_seconds // self.discretization.time_step
+        if self.runtime_code.working_path is None:
+            raise FileNotFoundError(
+                "local copy of ROMSSimulation.runtime_code does not exist. "
```
nitpick! missed a capital "L" to start the exception message

dafyddstephenson replied:
91884bc

Commit:
Horrible rabbit hole trying to fix roundtrip bug for ROMS analytical example cases:
- Made 'ininame' Optional[Path] in InitialConditions to allow for nrrec=0 and no ininame as in $ROMS_ROOT/Examples/Rivers_ana/river_ana.in
- Made 'filenames' Optional[list[Path]] in Forcing to allow for empty forcing section as in $ROMS_ROOT/Examples/Rivers_ana/river_ana.in
- add 'None' case to ROMSRuntimeSettingsSection.default_serializer for above cases
- Add custom 'from_lines' to InitialConditions and Forcing to handle above cases
- Add handling of 'Optional' in annotations in 'ROMSRuntimeSettingsSection.from_lines()'
- Realise '$ROMS_ROOT/src/read_inp.F' actually is case-sensitive and remove lower() from from_file/reset 's-coord -> 'S-coord' as in ROMS
- Decided once and for all that pydantic really isn't for me

dafyddstephenson commented:
> Notes to self: The roundtrip `ROMSRuntimeSettings.from_file(f).to_file(f2)` still fails for:
> 
>     * `$ROMS_ROOT/Examples/Rivers_ana/river_ana.in` (no initial condition file)
> 
>     * `$ROMS_ROOT/Examples/Pipes_ana/pipes_ana.in` (no initial condition file)
> 
>     * `$ROMS_ROOT/Examples/Filament/Filament.in` (no forcing files)
> 
> 
> Converting to draft until fixed

Ok fixed, took literally all day. Gross. Good to go after addressing review comments which are hopefully extremely gentle and easy to deal with so I never have to think about this module again ❤️

dafyddstephenson commented:
oh yeah, unit tests. Tomorrow it is.

Commit:
update unit tests

Commit:
add pydantic version to conda env

Commit:
PR review comment: https://github.com/CWorthy-ocean/C-Star/pull/245#discussion_r2095912496

Commit:
https://github.com/CWorthy-ocean/C-Star/pull/245#discussion_r2095913621

Commit:
remove unnecessary brackets

Commit:
https://github.com/CWorthy-ocean/C-Star/pull/245#discussion_r2096100425

Commit:
https://github.com/CWorthy-ocean/C-Star/pull/245#discussion_r2096112465

Commit:
https://github.com/CWorthy-ocean/C-Star/pull/245#discussion_r2096115454

Commit:
https://github.com/CWorthy-ocean/C-Star/pull/245#discussion_r2096120678

Commit:
https://github.com/CWorthy-ocean/C-Star/pull/245#discussion_r2096222878

Commit:
https://github.com/CWorthy-ocean/C-Star/pull/245#discussion_r2096344434

Commit:
https://github.com/CWorthy-ocean/C-Star/pull/245#discussion_r2096345120

Commit:
fix unit tests: cast_to_obj -> single_entry_validator

Commit:
https://github.com/CWorthy-ocean/C-Star/pull/245#discussion_r2096429040

Commit:
refactor load_raw_sections to avoid double while loop (https://github.com/CWorthy-ocean/C-Star/pull/245#discussion_r2096436772)

Commit:
https://github.com/CWorthy-ocean/C-Star/pull/245#discussion_r2096465834

Commit:
https://github.com/CWorthy-ocean/C-Star/pull/245#discussion_r2096454871

dafyddstephenson commented:
ok I think I've addressed as much as I can - @ankona @ScottEilerman good to go?

ScottEilerman commented:
> ok I think I've addressed as much as I can - @ankona @ScottEilerman good to go?

I didn't closely follow Chris's review, but yep, my approval still stands if everyone else is satisfied. Good work on this!

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+class ROMSRuntimeSettingsSection(BaseModel):
+    section_name: ClassVar[str]
```
Avoid unnecessarily verbose member naming. `Section.section_name` is redundant. Prefer `Section.name`

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+def _format_float(val) -> str:
+    if val == 0.0:
+        return "0."
+    elif abs(val) < 1e-2 or abs(val) >= 1e4:
```
Unnecessary `elif` branches after return statements

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+# Formatting methods for serializer:
+
+
+def _format_list_of_floats(float_list: list[float]) -> str:
```
missing docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+def _format_path(path: Path) -> str:
+    return str(path)
```
docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+def _format_other(other: str | int) -> str:
+    return str(other)
```
docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+    key_order: ClassVar[list[str]]
+
+    def __init__(self, *args, **kwargs):
+        if args:
```
docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+    @property
+    def value_joiner(self):
+        return "\n    " if self.multi_line else "    "
```
docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+    @model_serializer
+    def default_serializer(self) -> str:
+        # Make a dictionary out of the attributes:
```
docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+class Title(SingleEntryROMSRuntimeSettingsSection):
+    title: str
```
class docstring

ScottEilerman replied:
in lieu of future discussion, for now, I'm going to decree that "small" pydantic subschemas don't need docstrings.

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+class OutputRootName(SingleEntryROMSRuntimeSettingsSection):
+    output_root_name: str
```
class docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+class Rho0(SingleEntryROMSRuntimeSettingsSection):
+    rho0: float
```
class docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+class Gamma2(SingleEntryROMSRuntimeSettingsSection):
+    gamma2: float
```
class docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+class LateralVisc(SingleEntryROMSRuntimeSettingsSection):
+    lateral_visc: float
```
class docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+class TracerDiff2(SingleEntryROMSRuntimeSettingsSection):
+    tracer_diff2: list[float]
```
class docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+class SSSCorrection(SingleEntryROMSRuntimeSettingsSection):
+    sss_correction: float
```
class docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+class SSTCorrection(SingleEntryROMSRuntimeSettingsSection):
+    sst_correction: float
```
class docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+class UBind(SingleEntryROMSRuntimeSettingsSection):
+    ubind: float
```
class docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+class VSponge(SingleEntryROMSRuntimeSettingsSection):
+    v_sponge: float
```
class docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+class Grid(SingleEntryROMSRuntimeSettingsSection):
+    grid: Path
```
class docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+class Climatology(SingleEntryROMSRuntimeSettingsSection):
+    climatology: Path
```
class docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+class TimeStepping(ROMSRuntimeSettingsSection):
+    ntimes: int
```
class docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+class BottomDrag(ROMSRuntimeSettingsSection):
+    rdrg: float
```
class docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+class InitialConditions(ROMSRuntimeSettingsSection):
+    nrrec: int
```
class docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+class Forcing(ROMSRuntimeSettingsSection):
+    filenames: Optional[list[Path]]
```
class docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+class VerticalMixing(ROMSRuntimeSettingsSection):
+    Akv_bak: float
```
class docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+class MARBLBiogeochemistry(ROMSRuntimeSettingsSection):
+    marbl_namelist_fname: Path
```
class docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+class SCoord(ROMSRuntimeSettingsSection):
+    theta_s: float
```
class docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+class LinRhoEos(ROMSRuntimeSettingsSection):
+    Tcoef: float
```
class docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+class MYBakMixing(ROMSRuntimeSettingsSection):
+    Akq_bak: float
```
class docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+class ROMSRuntimeSettings(BaseModel):
+    # Non-optional:
```
class docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+## Class to hold all sections:
+
+
+class ROMSRuntimeSettings(BaseModel):
```
the "# non-optional" comment is unnecessary. we can see that from the typehints

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+    forcing: Forcing
+    output_root_name: OutputRootName
+    # Optional:
+    rho0: Optional[Rho0] = None
```
the "# optional" comment is unnecessary. we can see that from the typehints

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+    @staticmethod
+    def _load_raw_sections(filepath: Path) -> dict[str, list[str]]:
+        if not filepath.exists():
```
docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+        """
+
+        # Read file
+        filepath = Path(filepath)
```
appears to fail to validate the existence of the file before using it. this should raise a file-not-found error or similar to enable identifying the source of a failure.

Note: i think this would use "." if the filepath is an empty string. Sanity check your corner-case inputs

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+        )
+        lines.append("Time stepping (`ROMSRuntimeSettings.time_stepping`):")
+        lines.append(f"- Number of steps (`ntimes`) = {self.time_stepping.ntimes},")
+        lines.append(f"- Time step (`dt`, sec) = {self.time_stepping.dt},")
```
should stringification be delegated to the classes this object is composed of to avoid a requirement to update code in two places when the settings classes change?

ScottEilerman replied:
probably, but leaving to future work.

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+            "v_sponge": self.v_sponge,
+        }
+
+        inner = ", ".join(f"{k}={repr(v)}" for k, v in attrs.items() if v is not None)
```
consider using the formatting shortcut `f"{foo=}` where possible to avoid having to synchronize the hardcoded keys above and the actual property names on the objects.

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+        filepath = Path(filepath)
+
+        output_order = [
```
You should be able to rely on the order the properties are added to the model instead of hardcoding yet another list of names. re-order the properties in the model definition to meet your desired order, if possible.

If you're selectively printing attributes, consider using an exclude list to make the amount of hard-coding shorter (or use a decorator to mark printables, or use an annotation)

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+            "climatology",
+        ]
+
+        with filepath.open("w") as f:
```
The call to `getattr` here will hide errors we should be seeing. if the list is incorrect, implement at the right location in the class hierarchy to avoid hiding issues with the hard-coded list.

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+                if getattr(self, field) is None:
+                    continue
+                fieldname = self.model_fields[field].alias or field
+                f.write(self.model_dump(by_alias=True)[fieldname])
```
This method should be implemented as a custom serializer to enable using `pydantic` to do the heavy lifting. 

If you need a separate view of the object for this file, create another class to provide exactly that view. That would also avoid the hardcoded list of attribute names.

ScottEilerman replied:
I'm not sure we moved much of the heavy lifting, but I did split out the stringification to a custom serializer, separated from the to_file call.

ankona commented on cstar/roms/simulation.py:
```diff
@@ -240,16 +252,59 @@ def __init__(
+        """
+
+        in_files = [f for f in self.runtime_code.files if f.endswith(".in")]
+        if len(in_files) != 1:
```
consider two checks to improve our ability to solve the resulting issues.

```py
if not in_files:
   raise RuntimeError("No .in file was provided")

if len(in_files) > 1
   raise RuntimeError("1 file is allowed."
```

ankona commented on cstar/roms/simulation.py:
```diff
@@ -240,16 +252,59 @@ def __init__(
+            )
+        self._in_file = in_files[0]
+
     def _check_forcing_collection_types(self, collection, expected_class):
```
add type-hints to this method while we're adding the docstring

ankona commented on cstar/roms/simulation.py:
```diff
@@ -439,57 +514,165 @@ def codebases(self) -> list[ExternalCodeBase]:
+
+        for source in filter(None, forcing_sources):
+            paths = source.working_path
+            if paths is None:
```
check this outside the loop to simplify the loop itself.

```python
if any(source.working_path is None for source in forcing_sources):
    raise ValueError(" does not have wp")

for source in forcing_sources:
   ...
```

note: if the filter was implemented `filter(lambda x: x is not None, forcing_sources)` you could also avoid this check every time through the loop

ScottEilerman replied:
See below comment.

ankona commented on cstar/roms/simulation.py:
```diff
@@ -439,57 +514,165 @@ def codebases(self) -> list[ExternalCodeBase]:
+                )
+
+            elif isinstance(paths, list):
+                forcing_paths.extend(paths)
```
why are we doing these two checks? above, we purposefully inject a list:

```python
forcing_sources = chain(
            [self.tidal_forcing, self.river_forcing],
            self.surface_forcing,
            self.boundary_forcing,
            self.forcing_corrections,
        )
```

and here, we have to deal with it differently.

Can't we rewrite this method to avoid the loop completely?

```python
def _forcing_paths():
    forcing_sources = chain(
        self.tidal_forcing, 
        self.river_forcing,
        self.surface_forcing,
        self.boundary_forcing,
        self.forcing_corrections,
    )

    if any(source for source in forcing_sources if not source.working_path):
        raise ValueError(" does not have wp")

    return [source.working_path for source in forcing_sources]
   
```

ScottEilerman replied:
I think the wonkiness here is that some of the forcings can be list of strings/paths and others are single strings/paths, so the extra handling is to iterate/append/extend without accidentally iterating over a string instead of a list. Could be cleaned up, and I started taking a stab at it, but decided to leave it for now. The `for` loop helps to identify the particular item that has a problem, if there is one, as opposed to the `any()`. We might end up refactoring forcings in the coming month or two anyway to be more consistent.

ankona commented on cstar/roms/simulation.py:
```diff
@@ -439,57 +514,165 @@ def codebases(self) -> list[ExternalCodeBase]:
+        """
+
+        run_length_seconds = int((self.end_date - self.start_date).total_seconds())
+        return run_length_seconds // self.discretization.time_step
```
This will likely produce an incorrect number of time steps (off by 1). The remainder must be processed in an additional timestep.

add
```py
if run_len_seconds % self.d.ts:
  rls += 1
return rls
```

ScottEilerman replied:
I am guessing we want the floor division here to avoid running out of data on the last time step of the simulation. Will defer to @dafyddstephenson .

ankona commented on cstar/roms/simulation.py:
```diff
@@ -439,57 +514,165 @@ def codebases(self) -> list[ExternalCodeBase]:
+        # Time step entry
+        simulation_runtime_settings.time_stepping.dt = self.discretization.time_step
+
+        # ntimesteps entry:
```
i don't know if it's necessary to have a comment saying "i'm setting the X property" before setting the X property.

Ditto above for `time_stepping.dt`

ankona commented on cstar/roms/simulation.py:
```diff
@@ -439,57 +514,165 @@ def codebases(self) -> list[ExternalCodeBase]:
+            self.runtime_code.working_path / self._in_file
+        )
+
+        # Previous modifications
```
"Previous modifications" doesn't make any sense to me reading this the first time. What are you trying to say here. Do you need this comment? Can it be improved or removed?

ankona commented on cstar/roms/simulation.py:
```diff
@@ -439,57 +514,165 @@ def codebases(self) -> list[ExternalCodeBase]:
+        simulation_runtime_settings.time_stepping.ntimes = self._n_time_steps
+
+        # Initial conditions entry
+        if self.initial_conditions:
```
Unnecessary comment that is likely to go stale over time. we can see that we're editing initial_conditions. The comment should add context instead of re-state the code.

ankona commented on cstar/roms/simulation.py:
```diff
@@ -439,57 +514,165 @@ def codebases(self) -> list[ExternalCodeBase]:
-                + "ROMS expects a runtime options file with the '.in'"
-                + "extension, e.g. roms.in"
+
+        # Grid entry
```
Unnecessary comment `# Grid entry` that is likely to go stale over time. we can see that we're editing the grid. The comment should add context instead of re-state the code.

If you want to call out sections of behavior, a better approach is to create methods that can be called.

ankona commented on cstar/roms/simulation.py:
```diff
@@ -439,57 +514,165 @@ def codebases(self) -> list[ExternalCodeBase]:
-                return Path(in_files[0])
+            simulation_runtime_settings.grid = None
+
+        # Forcing
```
Ditto. unnecessary comment

ankona commented on cstar/roms/simulation.py:
```diff
@@ -439,57 +514,165 @@ def codebases(self) -> list[ExternalCodeBase]:
+                "marbl_diagnostic_output_list",
+            ]
+        ):
+            simulation_runtime_settings.marbl_biogeochemistry = (
```
this line of code is incomprehensible due to the formatter. this should be broken into individual variable sets instead of nesting and doing repeated line-continuations.

ankona commented on cstar/roms/simulation.py:
```diff
@@ -1408,14 +1389,19 @@ def run(
         if self.discretization.n_procs_tot is None:
             raise ValueError(
                 "Unable to calculate node distribution for this Simulation. "
                 + "Simulation.n_procs_tot is not set"
```
unnecessary string concatenations should be replaced with implicit concatenations (no need for `+`)

ankona commented on cstar/roms/simulation.py:
```diff
@@ -1408,14 +1389,19 @@ def run(
+        if self.runtime_code.working_path is None:
+            raise FileNotFoundError(
+                "Local copy of ROMSSimulation.runtime_code does not exist. "
+                + "Call ROMSSimulation.setup() or ROMSSimulation.runtime_code.get() "
```
unnecessary string concatenations should be replaced with implicit concatenations (no need for +)

ankona commented on cstar/roms/simulation.py:
```diff
@@ -1408,14 +1389,19 @@ def run(
-        n_time_steps = run_length_seconds // self.discretization.time_step
+        if self.runtime_code.working_path is None:
+            raise FileNotFoundError(
+                "Local copy of ROMSSimulation.runtime_code does not exist. "
```
Consider adding the failing path to the error message to enable debugging

ankona commented on cstar/roms/simulation.py:
```diff
@@ -1428,27 +1414,17 @@ def run(
         # these are conceptually different:
         run_path = output_dir
```
they might be conceptually different but they have the same value.... is this comment still valid? if they're equal, how are they conceptually different? Can we update or remove this while we're here

ankona commented on cstar/roms/simulation.py:
```diff
@@ -1428,27 +1414,17 @@ def run(
-            "__NTIMES_PLACEHOLDER__",
-            str(n_time_steps),
+        final_runtime_settings_file = (
+            self.runtime_code.working_path.resolve() / f"{self.name}.in"
```
Can we perform this resolve earlier (e.g. in a property) when working_path is set to avoid a need to resolve every time it's used?

ScottEilerman replied:
This is a change we should probably apply to all working paths -- will leave for future cleanup / consideration.

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+log = get_logger(__name__)
+
+################################################################################
```
if you want these to be identifiable as a set of related functionality, avoid comments and place in a module, instead.

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+
+def _format_list_of_other(other_list: list[str | int]) -> str:
+    joiner = " "
```
missing docstring

ankona replied:
rename `joiner` to `delimiter` to give var a meaningful name. a "joiner" is something that joins (performs the process) while a delimiter is the thing that goes between two things.

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+def _format_list_of_paths(
+    path_list: list[Path], multi_line: Optional[bool] = False
+) -> str:
+    joiner = "\n    " if multi_line else " "
```
1. missing docstring
2. the delimiter `"\n   "` is duplicated and should be placed into a constant to avoid requirement to change in two places later and to keep the spacing consistent (e.g. typos if adding another copy somewhere else later).

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+    return joiner.join(_format_other(x) for x in other_list)
+
+
+def _format_float(val) -> str:
```
missing docstring

ankona replied:
missing typehint on input argument `val`

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+    return str(other)
+
+
+################################################################################
```
see comment above. separate functionality into modules instead of using comments that can easily fall out of sync with content. remove.

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+################################################################################
+
+
+class ROMSRuntimeSettingsSection(BaseModel):
```
class docstring missing

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+class ROMSRuntimeSettingsSection(BaseModel):
+    section_name: ClassVar[str]
+    multi_line: ClassVar[bool] = False
+    key_order: ClassVar[list[str]]
```
members should have docstrings for the documentation and dev tools.

```py
multi_line: ...
"""Triple tic comment will show up in docs and IDEs"""
```

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+    multi_line: ClassVar[bool] = False
+    key_order: ClassVar[list[str]]
+
+    def __init__(self, *args, **kwargs):
```
missing typehints for args and return type

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+    def __init__(self, *args, **kwargs):
+        if args:
+            super().__init__(**{k: args[i] for i, k in enumerate(self.key_order)})
```
a hard-coded key order is too easy to fall out of sync with model changes. consider making these init methods keyword only so we can get rid of ordering

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+    @model_validator(mode="wrap")
+    @classmethod
+    def validate_from_lines(cls, data, handler: ModelWrapValidatorHandler):
+        # if the class gets a list of strings as it's init, assume it's coming in as a line
```
Convert to a real docstring

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+    @model_validator(mode="wrap")
+    @classmethod
+    def validate_from_lines(cls, data, handler: ModelWrapValidatorHandler):
```
Add typehint and return type hint (even if `t.Any`)

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+            try:
+                return cls.from_lines(data)
+            except Exception as e:
+                log.debug(
```
Consider passing the trace instead of manually adding type and method information:
`logger.debug(short_msg, stack_info=True)`

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+                        _format_list_of_paths(value, multi_line=self.multi_line)
+                    )
+                case list():
+                    section_values_as_list_of_str.append(_format_list_of_other(value))
```
nitpick: can we use a sane length for this variable name? `section_values` would do.

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+        # Make a dictionary out of the attributes:
+        data = {k: getattr(self, k) for k in self.key_order}
+
+        section_values_as_list_of_str = []
```
tighten type hinting w/`list[str]` instead of `list[any]`

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+                    section_values_as_list_of_str.append(_format_float(value))
+                case Path():
+                    section_values_as_list_of_str.append(_format_path(value))
+                case None:
```
filter `None` values prior to iteration to avoid having useless case

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+            section_values_as_list_of_str
+        )
+
+        section_header = f"{self.section_name}: {' '.join(data.keys())}\n"
```
any reason we're not using the `value_joiner` method to ensure consistent spacing?

ScottEilerman replied:
per above, this was intentional, though not spelled out as such.

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+        # Build the serialized string
+        serialized = ""
+        serialized += section_header
+        serialized += f"    {section_values_as_single_str}\n"
```
Is the whitespace here consistent with the whitespace elsewhere (e.g. `"\n   " if multiline else " "`)? Can we ensure consistency by re-using an indent constant instead.

ScottEilerman replied:
I'm not sure if this one is intentionally different, but at this point I'm leaving it.

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+    @classmethod
+    def from_lines(
+        cls, lines: Optional[list[str]]
```
It makes no sense to "encourage" a null parameter with `Optional` here.

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+
+        Parameters
+        ----------
+        lines: list[str]
```
typehint is not consistent with hardcoded docstring type

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+        >>> InitialConditions.from_lines(["1", "input_datasets/roms_ini.nc"])
+            InitialConditions(nrrec=1, ininame=Path("input_datasets/roms_ini.nc"))
+        """
+        if not lines:
```
this should raise an Exception. a file that should have content does not and the simulation should not be expected to behave correctly without the data.

ankona commented on cstar/roms/runtime_settings.py:
```diff
@@ -0,0 +1,806 @@
+        return handler(data)
+
+    def __init_subclass__(cls, **kwargs):
+        """Overrides default __init_subclass__ to allow definition of
```
decorate with `@overrides` to avoid stating relationship to another class and keep docstring talking about what it does.

ScottEilerman committed:
working with metaclass

ScottEilerman committed:
working w/o subclasses

ScottEilerman committed:
clean up aliases and some validation

ScottEilerman committed:
Lots of cleanup, PR comments, tests

ScottEilerman committed:
Remove testing block

ScottEilerman committed:
Align output with previous

ScottEilerman committed:
clean up from self-review

ScottEilerman committed:
Move persist call for local processing

ScottEilerman committed:
readiblity pass

ScottEilerman committed:
PR review

ScottEilerman committed:
Remove testing block that I put back in

ScottEilerman committed:
Used pydantic TypeAdapter instead of gnarly manual type-checking

dafyddstephenson committed:
Merge pull request #60 from ScottEilerman/roms_input_settings

Modifications for PR #245

ScottEilerman reviewed (commented):
We already merged this per internal conversations that it was good to move forward, but for posterity, I'll publish my response to some of the comments. I'm also going to make a separate issue about potentially raising an error if the timestep doesn't divide equally.
